### PR TITLE
Move logic from from_str to initializers

### PR DESCRIPTION
**Pull request by dexter2206**
## Description
 
It was noticed that `from_str` method of `PauliTerm` and `PauliSum` is so commonly used, that it might actually be better to just move the logic from there to the initializer. This PR implements this change, along with some minor improvements of logic and tests.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: src/orquestra/quantum/wip/operators/_pauli_operators.py [hunk 1]
```diff
@@ -371,8 +366,12 @@ def is_constant(self) -> bool:
 
 
 class PauliSum:
-    def __init__(self, terms: Sequence[PauliTerm] = None):
-        if terms is None:
+    def __init__(self, terms: Union[str, PauliTerm, Sequence[PauliTerm]] = None):
+        if isinstance(terms, str):
+            terms = [PauliTerm(s.strip()) for s in re.split(r"\+(?![^(]*\))", terms)]
+        elif isinstance(terms, PauliTerm):
```
Comment (laurgao) [inline src/orquestra/quantum/wip/operators/_pauli_operators.py:372]: Curious what's the intention behind adding this functionality to construct a PauliSum from a singular PauliTerm?

File: src/orquestra/quantum/wip/operators/_pauli_operators.py [hunk 2]
```diff
@@ -83,6 +84,34 @@ def _validate_type(object: Any) -> None:
...
+    if not match:
+        raise ValueError("Badly formatted string representation passed.")
+
+    return int(match.group(2)), match.group(1).upper()
+
+
+def _parse_operators_and_coefficient(
+    term_str: str,
+) -> Tuple[Optional[complex], Dict[int, str]]:
+    parts = term_str.split("*")
```
Comment (laurgao) [inline src/orquestra/quantum/wip/operators/_pauli_operators.py:99]: I would strip the string of all spaces at the very beginning. Right now it errors out when I pass a string such as `PauliTerm("2 * Z1 * Z2")`

File: src/orquestra/quantum/wip/operators/_pauli_operators.py [hunk 3]
```diff
@@ -91,40 +120,35 @@ class PauliTerm:
...
-                    "Invalid qubit index in dictionary keys."
-                    " Make sure all qubit indices are non-negative integers."
-                )
-
-            # Verify value values
-            if not all([op in ALLOWED_OPERATORS for op in operator.values()]):
+        if isinstance(operator, str):
+            _parsed_coefficient, operator = _parse_operators_and_coefficient(operator)
+            if _parsed_coefficient is not None and coefficient is not None:
                 raise ValueError(
```
Comment (laurgao) [inline src/orquestra/quantum/wip/operators/_pauli_operators.py:128]: maybe add a test that makes sure this ValueError happens when we pass a coefficient to both the string and argument

File: tests/orquestra/quantum/wip/operators/_operators_test.py [hunk 4]
```diff
@@ -80,13 +80,13 @@ def test_pauliterm_from_list_constructor(self, pauli_term):
         assert "Duplicate" in str(e.value)
 
     def test_pauliterm_from_str_constructor(self, pauli_term):
-        assert PauliTerm.from_str(str(pauli_term)) == pauli_term
+        assert PauliTerm(str(pauli_term)) == pauli_term
```
Comment (laurgao) [inline tests/orquestra/quantum/wip/operators/_operators_test.py:83]: I would also add some test cases of different representations such as placing the coefficient in the argument vs the string, having spaces in the string, etc. And also some more badly formatted string representations such as incorrect bracket placement around the operators. Thoughts?

**Review by laurgao — COMMENTED** [review]
After we're done with this PR, we'll also have to update the changes in orquestra-opt to not use the `from_str` method.

File: src/orquestra/quantum/wip/operators/_pauli_operators.py (hunk 1 above)
Comment (dexter2206) [reply src/orquestra/quantum/wip/operators/_pauli_operators.py]: You're right, I removed it, we'll add it again if it turns out to be needed.

File: src/orquestra/quantum/wip/operators/_pauli_operators.py (hunk 2 above)
Comment (dexter2206) [reply src/orquestra/quantum/wip/operators/_pauli_operators.py]: It should no longer error out for strings with spaces inside them.

File: src/orquestra/quantum/wip/operators/_pauli_operators.py (hunk 3 above)
Comment (dexter2206) [reply src/orquestra/quantum/wip/operators/_pauli_operators.py:140]: Suitable test has been added.

File: tests/orquestra/quantum/wip/operators/_operators_test.py (hunk 4 above)
Comment (dexter2206) [reply tests/orquestra/quantum/wip/operators/_operators_test.py]: I added more test cases and split them into more test functions for clarity. Let me know wyt :)

File: tests/orquestra/quantum/wip/operators/_operators_test.py [hunk 5]
```diff
@@ -4,149 +4,286 @@
...
+        # from_list accepts input in the reverse direction then the
+        # operations_as_set produces it.
+        assert term.operations_as_set() == frozenset([pair[::-1] for pair in op_list])
+        assert term.coefficient == coefficient
+
+    def test_term_cannot_be_constructed_from_list_of_incorrectly_shaped_tuples(self):
         with pytest.raises(ValueError) as e:
             PauliTerm.from_list([("X0")])
+
         assert "PauliTerm.from_str" in str(e.value)
```
Comment (laurgao) [inline tests/orquestra/quantum/wip/operators/_operators_test.py:169]: this error message should be changed as we are no longer using the `PauliTerm.from_str` method

File: src/orquestra/quantum/wip/operators/_pauli_operators.py [hunk 6]
```diff
@@ -83,58 +84,94 @@ def _validate_type(object: Any) -> None:
...
-    """
-    A datastructure for storing information about a single Pauli Term
-    """
+    """Representation of a single Pauli Term."""
 
     def __init__(
         self,
         operator: Union[str, Dict[int, str]],
-        coefficient: complex = 1.0,
+        coefficient: Optional[complex] = None,
```
Comment (laurgao) [inline src/orquestra/quantum/wip/operators/_pauli_operators.py:135]: [nitpick] maybe add a note that the default coefficient is 1 now that it's no longer clear from reading the default arguments?

**Review by laurgao — APPROVED** [review]

File: tests/orquestra/quantum/wip/operators/_operators_test.py (hunk 4 above)
Comment (laurgao) [reply tests/orquestra/quantum/wip/operators/_operators_test.py]: Tests look great!

File: src/orquestra/quantum/wip/operators/_pauli_operators.py (hunk 6 above)
Comment (dexter2206) [reply src/orquestra/quantum/wip/operators/_pauli_operators.py:139]: Added in [c8c6056](https://github.com/zapatacomputing/orquestra-quantum/pull/44/commits/c8c605691a35ff4232d0d6c42c5690e3c9e859c1)

File: tests/orquestra/quantum/wip/operators/_operators_test.py [hunk 7]
```diff
@@ -318,12 +455,8 @@ def test_paulisum_native_constructor(self):
         with pytest.raises(ValueError):
             PauliSum([PauliTerm.identity(), 1])
 
-        # Non-list input
-        with pytest.raises(ValueError):
-            PauliSum(PauliTerm.identity())
```
[the remark this answers is not in the document]
Comment (dexter2206) [reply tests/orquestra/quantum/wip/operators/_operators_test.py:323]: Changed in [fb15052](https://github.com/zapatacomputing/orquestra-quantum/pull/44/commits/fb150526e64eb114cad13752e57c23c6fb707814) along with some fixes in logic how the constructor checks if non-list is provided.

File: tests/orquestra/quantum/wip/operators/_operators_test.py (hunk 5 above)
Comment (dexter2206) [reply tests/orquestra/quantum/wip/operators/_operators_test.py]: Changed in [6e4fac1](https://github.com/zapatacomputing/orquestra-quantum/pull/44/commits/6e4fac19fc25b642d4e66f23c45fc31e9e18fc03)

**Pull request merged**